### ISSUE TITLE
Proposal: add snapshot testing (this is not screenshot diffing)

### ISSUE THREAD

**Issue by bahmutov** [cypress-io/cypress#772]
Similar to Jest / Ava snapshot testing (see https://glebbahmutov.com/blog/snapshot-testing/ and https://glebbahmutov.com/blog/picking-snapshot-library/) it would be nice to save / compare values to snapshots.

For example this test can save the wrapped objects on the first run, then we would check in the generated `snapshots.js` file. Next time Cypress runs it would compare new values with saved snapshots and throw an error if they are different.

Here is a proof of concept

```js
describe('my tests', () => {
  it('works', () => {
    cy.log('first snapshot')
    cy.wrap({foo: 42}).snapshot()

    cy.log('second snapshot')
    cy.wrap({bar: 101}).snapshot()
  })
})
```
The test runner

![screen shot 2017-10-19 at 1 52 55 am](https://user-images.githubusercontent.com/2212006/31756219-d9a24b8e-b470-11e7-9260-dcdc64c43cc1.png)

When clicking on "snapshot" in the reporter, it shows the object in the dev tools

Similarly, if previous command yields a jQuery element, then it is converted into JSON and saved. For example, we could snapshot focused element before and after typing (TodoMVC example at todomvc.com/examples/react/)

```js
describe('focused input field', () => {
  it('is empty and then typed into', () => {
    cy.visit('http://todomvc.com/examples/react/')
    cy.focused()
      .snapshot('initial') // given human names as well
      .type('eat healthy breakfast')
      .snapshot('after typing')
  })
})
```
When clicking on the snapshot in this case, prints original element and JSON object

![screen shot 2017-10-19 at 1 53 07 am](https://user-images.githubusercontent.com/2212006/31756261-219ee852-b471-11e7-93d3-44a326d1db28.png)

The snapshots are all loaded at the start and saved together at the end. Here is snapshot file
```js
module.exports = {
  "my tests": {
    "works": {
      "1": {
        "foo": 42
      },
      "2": {
        "bar": 101
      }
    }
  },
  "focused input field": {
    "is empty and then typed into": {
      "initial": {
        "tagName": "input",
        "attributes": {
          "class": "new-todo",
          "placeholder": "What needs to be done?",
          "value": ""
        }
      },
      "after typing": {
        "tagName": "input",
        "attributes": {
          "class": "new-todo",
          "placeholder": "What needs to be done?",
          "value": "eat healthy breakfast"
        }
      }
    }
  },
  "__version": "1.0.2"
}
```

If a run time value during test does not match value loaded from snapshot, it throws an exception.
For example, changed code to `cy.wrap({foo: 45}).snapshot()` leads to

<img width="449" alt="screen shot 2017-10-19 at 2 01 27 am" src="https://user-images.githubusercontent.com/2212006/31756325-76138f28-b471-11e7-9bec-2bfbaa2601a6.png">

**Comment by brian-mann** [comment]
Loving this. Can we also add it for elements? 

```js
cy.get('form').snapshot() // should serialize the HTML as a string and inline certain things like value, checked, selected, etc
```

Bonus points if you format the HTML nicely 👍

**Comment by bahmutov** [comment]
It is already there - the focused().snapshot() is an example. Instead of HTML it converts DOM to json though

**Comment by brian-mann** [comment]
I see... is this what jest does as opposed to string HTML?

**Comment by bahmutov** [comment]
I think Jest converts to HTML, but I am not sure. But I think this is great user space feature, so I can make this public under `@bahmutov/cypress-snapshot`, see how it goes and then we can move into `@cypress/<..>` or build into Cypress itself.
Plus we can have options how to stringify values before comparing (already snapshot supports it under the hood), which is something Jest does too.

**Comment by brian-mann** [comment]
I want to see this in cypress core... it's incredibly useful and powerful for a myriad of things.

Using the console as a comparison is super nice to explore the object structures.

We could even potentially compare the HTML using the actual application under test by inlining all of the styles or capturing them as well separately and applying the styles to the elements.

**Comment by bahmutov** [comment]
Yes!!!

**Comment by jennifer-shehane** [comment]
Just wanted to bring up that we use the terminology 'snapshot' for explaining the 'time-travel' process of restoring the DOM when hovering over commands and 'pinning' - we will have to rename this to something else, since I think most people define 'snapshots' as the feature explained in this proposal.

**Comment by bondz** [comment]
Yes @jennifer-shehane. New to Cypress, and every time I read snapshot in the docs, I thought it had support for jest snapshots, which I thought was awesome. Seeing that is uses chai is what led me here.

**Comment by newtack** [comment]
We use Jest for the rest of our testing, so if Cypress could integrate with Jest that would be awesome.

**Comment by vinayakkulkarni** [comment]
We use Ava across our applications. Any Cy + ava examples?

**Comment by gsouf** [comment]
Interestingly there doesn't seem to have any good solution for json/data snapshots that is maintained for cypress 10+ and cypress mentioned 5 years ago that it's a plan to get it to the core https://www.cypress.io/blog/2018/01/16/end-to-end-snapshot-testing/.

**Comment by fcuenya** [comment]
Any progress on this, or if it has been replaced with a 3rd party plugin or a different approach, what would that be?

I want to incorporate this testing approach to my project and I was wondering what would be the latest best practices around it.
